payjoin/src/send/error.rs

```diff
@@ -235,3 +236,179 @@ impl std::error::Error for CreateRequestError {
+    Unrecognized(String, String),
+    /// `Validation` errors are errors that are caused by malformed responses.
+    /// They are only displayed in debug logs.
+    Validation(ValidationError),
+}
+
+impl ResponseError {
+    fn from_json(json: serde_json::Value) -> Self {
+        // we try to find the errorCode field and
+        // if it exists we try to parse it as a well known error
+        // if its an unknown error we return the error code and message
+        // from original response
+        // if errorCode field doesn't exist we return parse error
+        match json
+            .as_object()
+            .and_then(|v| v.get("errorCode"))
+            .and_then(|v| v.as_str())
+            .ok_or(InternalValidationError::Parse)
+        {
+            Ok(str) =>
+                if let Ok(known_error) = WellKnownError::from_str(str) {
+                    return known_error.into();
+                } else {
+                    let message = json
+                        .as_object()
+                        .and_then(|v| v.get("message"))
+                        .and_then(|v| v.as_str())
+                        .ok_or(InternalValidationError::Parse)
+                        .unwrap_or("");
+                    return Self::Unrecognized(str.to_string(), message.to_string());
+                },
+            Err(e) => return e.into(),
+        }
+    }
+    /// Parse a response from the receiver.
+    ///
+    /// response must be valid JSON string.
+    pub fn from_str(response: &str) -> Self {
+        if let Some(parsed) = serde_json::from_str(response).ok() {
+            Self::from_json(parsed)
+        } else {
+            InternalValidationError::Parse.into()
+        }
+    }
+}
+
+impl std::error::Error for ResponseError {}
+
+impl From<WellKnownError> for ResponseError {
+    fn from(value: WellKnownError) -> Self { Self::WellKnown(value) }
+}
+
+impl From<InternalValidationError> for ResponseError {
+    fn from(value: InternalValidationError) -> Self {
+        Self::Validation(ValidationError { internal: value })
+    }
+}
+
+// It is imperative to carefully display pre-defined messages to end users and the details in debug.
+impl Display for ResponseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::WellKnown(e) => e.fmt(f),
+            // Don't display unknowns to end users, only debug logs
+            Self::Unrecognized(_, _) => write!(f, "The receiver sent an unrecognized error."),
+            Self::Validation(e) => write!(f, "The receiver sent an invalid response: {}", e),
+        }
+    }
+}
+
+impl fmt::Debug for ResponseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::WellKnown(e) => write!(
+                f,
+                r#"Well known error: {{ "errorCode": "{}",
+                "message": "{}" }}"#,
+                e.error_code(),
+                e.to_string(),
```

Comment:
we should print the saved json error message here and use to_string just for display